manual/development-process/test-accessibility.md

```diff
@@ -0,0 +1,99 @@
+[wcag]: https://www.w3.org/WAI/standards-guidelines/wcag/
+
+## What to test
+
+Some of these tests are subjective. For example, what is the most obvious and
+logical tab order? If you are not sure about anything, talk to the rest of the
+team to get other people's opinions, and to check whether it's something that
+we should test with users.
+
+### Manual tests
+
+* Zoom into 400% and check the page is still readable and functional
+* Disable images and styles (search for how to do this in your specific browser)
+  and check that the page is still readable and functional
+* Check all functionality works using only the keyboard, and that you don't get
+  stuck when tabbing through elements
+* Check that it's obvious which element has focus when tabbing to it with the
+  keyboard
+* Check that keyboard tab order is obvious
+* Check that semantic markup is used wherever possible (e.g. `<ul>` rather than
+  `<div>` for a list, or `<h1>` rather than a styled `<span>` for a header)
+
+### Screen reader tests
+
+Navigate through the page using a screen reader. Choose one which works on your
+operating system:
+
+* [VoiceOver] for MacOS
+* [NVDA] for Windows
+* [Orca] for Linux
+
+What to test with with the screen reader:
+
+* Check that the screen reader reads every element and header
+* Check that every link makes sense in the screen reader
+* Check that [ARIA] roles are being used where appropriate
+* Check that and that you can fill in every editable field
+* Check that whenever an element appears or is updated dynamically, the screen
+  reader announces the update in a way that is understandable
+
+Ideally, we would also test every new page with a speech recognition tool like
+[Dragon], but we don't currently have a licence.
+
+[VoiceOver]: https://webaim.org/articles/voiceover/
+[NVDA]: https://www.nvaccess.org/
+[Orca]: https://help.gnome.org/users/orca/stable/
+[ARIA]: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA
+[Dragon]: https://www.nuance.com/en-gb/dragon.html
+
+### Automated tests
+
+* Run either the [Chrome Lighthouse tool][lighthouse] or the [Firefox
+Accessibility Inspector][firefox-ai], and review any
+findings
+* Run the [WebAIM WAVE browser extension][wave] on the new page, and review any
+findings
+
+[lighthouse]: https://developers.google.com/web/tools/lighthouse
+[firefox-ai]: https://developer.mozilla.org/en-US/docs/Tools/Accessibility_inspector
+[wave]: https://wave.webaim.org/
+
+## Who should run the tests?
+
+Anyone on the team can run these tests, but it should normally be one of the
+developers who built the feature. This is to encourage all developers to get a
+deeper understanding of accessibility. Also, it's often easier and quicker to
+just fix something when you find an issue rather than reporting the problem to
+someone else.
+
+## When should we run the tests?
+
+We should always go through the full checklist once the feature has been
```

Comment:
Does this mean that a feature needs to be accessibility tested when it is deployed to intg, then again when it is deployed to staging, and again on prod deployment? If so, is that necessary, given the environments should all be the same?

Reply:
Nope, my wording was ambiguous, sorry. I meant that we should test it in _one_ of the environments, rather than localhost.

I think staging is ideal because it's a more stable environment than integration, but intg is probably fine in most circumstances. I think it's more critical to do end-to-end exploratory testing on staging, to avoid being tripped up by temporary changes on integration. But I think it's a bit less critical for accessibility testing because you're testing code that's focused in one application (normally the frontend, sometimes Keycloak).

Does my new wording make sense? I haven't got into the question of when we deploy to production, because I think we're going to have to build up norms around production deployment once the Beta is live. We can use techniques like feature flags to hide new features until they've been fully tested.